apps/sms/views/conversation/js/conversation.js

```diff
@@ -375,6 +375,34 @@ var ConversationView = {
     this.toggleRecipientSuggestions();
   },
 
+  hasValidRecipients() {
+    if (this.recipients.numbers.length) {
+      return true;
+    }
+
+    var notValidatedNumber = this.recipients.inputValue;
+    return notValidatedNumber && isFinite(notValidatedNumber);
+  },
+
+  hasEmailRecipients() {
+    if (!Settings.supportEmailRecipient) {
+      return false;
+    }
+
+    // When non-contact recipient is tapped by user, Recipients component
+    // removes recipient pill, fires "remove" event and only then populates
+    // recipients input with the content to edit. In case we have only one
+    // recipient, then in "remove" handler we'll have both empty recipients list
+    // and input (since input is not populated yet). At the same time in
+    // Conversation view we listen for the modification of recipients input. So
+    // we have two consequent events where we want to know how many and what
+    // recipients we have to possible disable send button and change message
+    // type if we have email recipient. That leads to two consequent message
+    // type change banners. To preserve original behavior and avoid double
+    // banner issue we just don't try to detect email in recipients input.
+    return this.recipients.numbers.some(Utils.isEmailAddress);
```

Comment:
please put the explanation between /* */ as it's lengthy :)

also maybe add new lines in some good places